tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Kafka/ProducerCache.cs

```diff
@@ -23,11 +27,43 @@ public static void AddProducer(object producer, string bootstrapServers)
 #endif
     }
 
+    public static void CreateDefaultDeliveryHandler<TProducer>(TProducer producer)
+    {
+        if (producer == null)
+        {
+            return;
+        }
+
+        var producerType = producer.GetType();
+        var deliveryReportType = producerType.Assembly.GetType("Confluent.Kafka.DeliveryReport`2");
+        var genericArgs = producerType.GetGenericArguments();
+
+        if (deliveryReportType != null)
+        {
+            var genericDeliveryReportType = deliveryReportType.MakeGenericType(genericArgs);
+            var delegateType = typeof(Action<>).MakeGenericType(genericDeliveryReportType);
```

Comment:
Yes, in general a Generic method that creates new types (that depends on the generic) should be promoted to a generic type, to do caching over those generic types in case it's called multiple times.

Reply:
A generic type will not work in this case. The reason for this is that `TTarget instance` for the instrumentation gets passed as `Object`, so `typeof(TTarget)` will give incorrect type. 
I know that using instance to cache will result in more handlers being created, however, there are typically a few producers per application, so the overhead should not be that bad.